httpclient5/src/main/java/org/apache/hc/client5/http/ssl/DefaultHostnameVerifier.java

```diff
@@ -114,18 +114,15 @@ public void verify(
                     matchIPv6Address(host, subjectAlts);
                     break;
                 default:
-                    matchDNSName(host, subjectAlts, this.publicSuffixMatcher);
+                    // In case there are no SubjectName.DNS entries, fallback to CN matching
```

Comment:
@peterdettman I am not sure this is what the RFC says. I believe that CN matching is to be used _only_ if there are no `dNSName` entries present.

Reply:
@peterdettman I only expected changes in handling of URIs specified as an IP address rather than a hostname.

Reply:
Yes, matchDNSNames is only returning false if there were no dNSName at all. If there were any present that didn’t match, it would have already thrown an exception.

Reply:
@peterdettman Then what is the point in calling `#matchCN` here?

Reply:
That is the fallback if there were no dNSName entries to even check against. RFC 2818 3.1: “If a subjectAltName extension of type dNSName is present, that MUST be used as the identity. Otherwise, the (most specific) Common Name field in the Subject field of the certificate MUST be used.”. Note that IP addresses are singled out further down: “In some cases, the URI is specified as an IP address rather than a hostname. In this case, the iPAddress subjectAltName must be present in the certificate and must exactly match the IP in the URI.”.

Reply:
Perhaps the code could be clearer if we only collected either IP or DNS entries depending on the HostNameType?

Reply:
For comparison, SunJSSE uses [HostnameChecker](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/sun/security/util/HostnameChecker.java) for the equivalent checks when endpoint identification is set to HTTPS. Likewise our (BCJSSE provider) uses [HostnameUtil](https://github.com/bcgit/bc-java/blob/master/tls/src/main/java/org/bouncycastle/jsse/provider/HostnameUtil.java).

Reply:
> Perhaps the code could be clearer if we only collected either IP or DNS entries depending on the HostNameType?

@peterdettman This is what I was trying to hint at. I really do not understand why there should be multiple invocations of `#matchCN`. Please refactor the code to make the execution flow as well as the intent of your changes a bit easier to understand.